### PR TITLE
Add spears mod

### PR DESCRIPTION
**Pull request by AKryukov92**
As suggested in #214 and #223.

### CODE REVIEW

**Comment by AKryukov92** [comment]
I don't like the idea of "weapons from all materials". That's why only stone and mese.

**Comment by LoneWolfHT** [comment]
I would use both a spear and sword when possible. I would probably prefer the sword. 1 node isn't much in PvP; Which goes all over the place. 
Spears of all materials would fit with the current tool materials setup

**Comment by AKryukov92** [comment]
You haven't played in CTF with spears at all. How do you know, that you will use both spear and sword?

**Comment by LoneWolfHT** [comment]
I've played LOTT. The sword is better for mobile PVP. The spear is good for stationary fighting. Like stabbing enemies on the ground while standing on a wall.

**Comment by AKryukov92** [comment]
@ClobberXD , do I understand you correctly: I should add steel and diamond spear to list of weapons to get approve on this PR?

**Comment by AKryukov92** [comment]
You have been granted write access for a reason. May be rubenwardy expects that you could help PRs. This fact worth clarification.

**Comment by ClobberXD** [comment]
I do not have write access to the master repo, and subsequently, can't merge PRs. But with the limited write-access I have, I can label issues and PRs to my heart's content. To put it in a more polished way - I help with triaging issues and PRs.

**Comment by AKryukov92** [comment]
Actually, no. I thought that treasure loot is defined in CTF-maps mod.
I'll add them then.

File: mods/spears/init.lua [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	range = 5,
+	tool_capabilities = {
+		full_punch_interval = 1.2,
+		max_drop_level=1,
+		groupcaps={
+			snappy={times={[1]=2.2, [2]=1.2, [3]=0.20}, uses=20, maxlevel=1},
+		},
+		damage_groups = {fleshy=4},
+	}
+})
```
[the remark this answers is not in the document]
Comment (AKryukov92) [reply mods/spears/init.lua]: Is it really an issue?

[the remark this answers is not in the document]
Comment (ClobberXD) [reply mods/spears/init.lua]: It's standard practice. If you're contributing to someone else's repository, you might as well contribute properly.

File: mods/ctf_crafting/depends.txt [hunk 2]
```diff
@@ -1 +1,2 @@
 crafting
+spears
```
[the remark this answers is not in the document]
Comment (ClobberXD) [reply mods/ctf_crafting/depends.txt]: That's interesting and unexpected. I wonder how other mods like `default`, `doors`, and `shooter` register their recipes correctly without having `ctf_crafting` depend on them... :thinking:

**Comment by ClobberXD** [comment]
If textures are taken from elsewhere, mentioning in the license and the PR's description is sufficient. While the textures would still work without renaming, it's recommended to follow this naming convention as documented in [lua_api.txt](https://github.com/minetest/minetest/blob/2e37ee956530b7cd310e0d6cfb8ce3cd8f6973e6/doc/lua_api.txt#L358):
```
Mods should generally prefix their textures with `modname_`, e.g. given
the mod name `foomod`, a texture could be called:

    foomod_foothing.png
```

There can be exceptions to this rule though: A mod might **directly use** a texture from another mod without copying (e.g. `boats` mod, that uses `default_wood.png` provided by `default`). In such cases, the aforementioned naming convention isn't followed, but is allowed. e.g. If `lottweapons` was a part of CTF and let's say you're adding a `spears` mod that uses textures from `lottweapons` directly, then this would be allowed.

File: mods/spears/init.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (AKryukov92) [reply mods/spears/init.lua]: Rubenwardy uses luacheck to force codestyle. That's why pull requests within trailing spaces can not be merged.
Luacheck do not require newlines at the end of the file. That's why I won't bother with it without any real reasons.

[the remark this answers is not in the document]
Comment (rubenwardy) [reply mods/spears/init.lua]: Please add the trailing new line. Luacheck does not enforce code style, it just looks for bugs and typos

File: mods/ctf_crafting/depends.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ClobberXD) [reply mods/ctf_crafting/depends.txt]: I just noticed this part - `and spears mod is missing`. Why would `spears` be missing if it's bundled with the game and can't be disabled? To be clear, there's nothing wrong with adding this, but it's just not worth it.

**Comment by ClobberXD** [comment]
> If stone spears will be used more often than stone swords, then steel spears will make steel swords useless.

This does not make much sense. It's like saying using swords will make pickaxes useless. While both spears and swords are used for PvP, both have their own unique properties and quirks. The mod should provide as many choices as possible. Using spears more than swords will vary from player to player, and it's their choice (and right) to do so.

Spears are generally known for their relatively higher range. I suggest increasing their range to 6 to better distinguish them from swords.

Will approve once the requested changes have been attended to.

**Comment by ClobberXD** [comment]
> In #223 I reduced range of swords to 3. If spears would have range equal to 5 like in this PR, then they will actually be very distinctive.

Oh, I see. So spears do have a range of two more nodes, then.

> I don't exactly know how stone spears will actually perform in the game. I think it is better to make underpowered weapon, than overpowered.

As I said, the main tactical advantage of spears is their relatively higher reach. The damage they do to players should be appropriately lowered, in order to not use spears in place of swords altogether.

**Comment by AKryukov92** [comment]
Assuming #223 , spears damage will be equal to pickaxe damage.

**Comment by ClobberXD** [comment]
> Assuming #223 , spears damage will be equal to pickaxe damage.

It'll be pretty well balanced, then. I don't see what's stopping you from adding spears of all materials. :)

File: mods/spears/init.lua [hunk 3]
```diff
@@ -1,10 +1,24 @@
 --Spears from mod Lottweapons, originally made by Amaz
 
 --Spears:
+minetest.register_tool("spears:spear_wood", {
```
Comment (ClobberXD) [inline mods/spears/init.lua]: Wooden swords just exist to be given as default items to players. I doubt anybody would bother to make wooden spears, considering that it'd be even weaker than a wooden sword. I don't think this is needed, but you could still keep it if you want, for the sake of completeness.

**Comment by ClobberXD** [comment]
Thank you for adding all variants of spears. I'll try to test them today.

File: mods/spears/init.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (AKryukov92) [reply mods/spears/init.lua]: Done

File: mods/ctf_crafting/depends.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (AKryukov92) [reply mods/ctf_crafting/depends.txt]: I'll keep `spears` in the dependency list, if you don't mind. It looks *right* way to do things.

**Comment by neinwhal** [comment]
something that might be worthwhile doing to balance it more, is increasing the full punch interval maybe by double, so spear can be made into a long range, but slow and decent damage weapon. while swords are short range, fast and deadly.

**Comment by LoneWolfHT** [comment]
> This PR is my proposal about properties of the spears.
> If you want different properties, then make your own different PR.
> I will change this PR only if repository admins will give me some important reasons to change it.

It would be a good idea to take into account what the players think. They are the ones that will be using this after all

**Comment by AKryukov92** [comment]
> It would be a good idea to take into account what the players think. They are the ones that will be using this after all

You are right. But PR is not a place for throwing ideas like "throw away your work and do as I said". Ideas should be discussed in issues. PR is a piece of work that is already done and proposed as is.

In addition, I don't think that anything can be balanced without gathering any statistics about actual usage of weapons. I will consider other possibilities only when this PR gets merged and community do some real playtesting.

**Comment by LoneWolfHT** [comment]
> > It would be a good idea to take into account what the players think. They are the ones that will be using this after all
> 
> You are right. But PR is not a place for throwing ideas like "throw away your work and do as I said". Ideas should be discussed in issues. PR is a piece of work that is already done and proposed as is.
> 
> In addition, I don't think that anything can be balanced without gathering any statistics about actual usage of weapons. I will consider other possibilities only when this PR gets merged and community do some real playtesting.

image

And this PR will never be merged if the admins see too much negative feedback. So either way your work would be rendered a waste of time.

**Comment by AKryukov92** [comment]
If this PR is not going to be merged, then it should be closed. Then, discussion should be continued in issues.
What is the point of having hanging PRs at all?

**Comment by AKryukov92** [comment]
If every random GitHub user will be able to influence content of the PR, then we end up in a mess.
There is still a possibility, that I actually change it. But they need more detailed explanation of reasons to change. Something more than just "I think you should change it".

**Comment by LoneWolfHT** [comment]
> If every random GitHub user will be able to influence content of the PR, then we end up in a mess.
> There is still a possibility, that I actually change it. But they need more detailed explanation of reasons to change. Something more than just "I think you should change it".

It might be a better idea to just ask for a more detailed explanation next time.

**Comment by ClobberXD** [comment]
> But PR is not a place for throwing ideas like "throw away your work and do as I said". Ideas should be discussed in issues.

Why would someone discuss about an open PR in a separate issue? Issues are for existing code. Discussions about PRs should take place in the PR page itself, which is the primary reason people are allowed to comment on a PR page.

> PR is a piece of work that is already done and proposed as is.

A PR is never "already done". It's supposed to be constantly improved through multiple discussions, opinions, and testing. Don't get me wrong, you have all the rights to reject others' opinions and inputs, but I'm just saying that it might probably result in your PR being rejected as well.

> If every random GitHub user will be able to influence content of the PR, then we end up in a mess.

The sole purpose of an open-source project is collaboration. Every random user is _supposed_ to provide (constructive) opinions/criticism, assuming they know what they're saying. Even if someone points out a trivial grammar mistake, that's only for the good of the PR, and hence for the good of the repository as well.

TL;DR: Being open to discussion and (constructive) criticism/opinions is always a good thing, as it will ensure that the PR is of good quality.

**Comment by AKryukov92** [comment]
> Don't get me wrong, you have all the rights to reject others' opinions and inputs, but I'm just saying that it might probably result in your PR being rejected as well.

Rejection of PR only because I reject (opinion without arguments) sounds like punishment for not being good for everyone. You can always punish me for not being nice to passerby, but I think it will not be constructive. @neinwhal did nothing to make his opinion matter. Why should I listen to every random github user?

"PR is constantly improving" is an attitude, which lead you to endless discussions without any result.
Can you tell me, when exactly discussion is supposed to end?
Can you describe quality criteria that should be satisfied to make this PR merged?

**Comment by ClobberXD** [comment]
> Rejection of PR only because I reject (opinion without arguments) sounds like punishment for not being good for everyone.

It isn't. I meant that it might probably be rejected for being low quality. If you aren't open to inputs and suggestions, the PR will **remain** low quality, resulting in its closure.

> Can you tell me, when exactly discussion is supposed to end?

Very simple. The discussion ends when there are no more suggestions being put forward, or the "repository admins" are completely satisfied with the PR. It also ends when the requested changes aren't being implemented by the author of the PR, for whatever reason. ;)

**Comment by AKryukov92** [comment]
> when there are no more suggestions being put forward

It means "never".
> "repository admins" are completely satisfied with the PR

It means nothing. Can you be more specific about satisfaction of repository admins? Should I send him a pizza for each PR?

> It also ends when the requested changes aren't being implemented by the author of the PR, for whatever reason

Are you sure, that every requested change should be implemented?
What if I come to your medkit PR and request recipe to make medkits from steel ingots? Will you implement it?

**Comment by ClobberXD** [comment]
> Should I send him a pizza for each PR?

I said "satisfied with the PR", not "satisfied with you".

> Are you sure, that every requested change should be implemented?

"Requested changes" are those requested by the "repository admins" via normal comments, or GitHub's code-review feature. Sorry for not clarifying that.

> What if I come to your medkit PR and request recipe to make medkits from steel ingots? Will you implement it?

I'll have a **discussion** with you about that suggestion not being meaningful, instead of asking you to go make another PR of your own. :)

**Comment by ClobberXD** [comment]
Although we're veering off-topic, I do agree on having some sort of an approval mechanism during the feature-request stage itself, so that potential contributors don't waste their time working on something that might be rejected right away. I hope @rubenwardy is more vocal with his opinion when it comes to feature requests.

**Comment by AKryukov92** [comment]
I am glad to see that we get understanding about approval mechanism of feature-requests.
But my point was:
PR should implement feature with specific description.
Feature description should not change during development.
PR should not change because feature has been changed.
If there is need of change, then another PR should be started.
In the end, repository admin should get multiple ready PRs and will be able to decide which should be merged and which should be rejected.

**Comment by ClobberXD** [comment]
> PR should implement feature with specific description.
Feature description should not change during development.
PR should not change because feature has been changed.
If there is need of change, then another PR should be started.

That'd result in too many branches with similar commits and unnecessary disconnect of conversation. i.e. The implementation discussions about the same feature will take place in multiple places, instead of one consolidated place.

**Comment by AKryukov92** [comment]
> unnecessary disconnect of conversation

If goal "Keep conversation going" is important, then you end up with endless conversation.
Contributors true goal is to improve the game. If we keep this in mind, then the game will be improved.

There is distinction between **implementation** discussions and **feature** discussions.
As I said earlier. implementation discussion should be in PR. It includes code style, optimisation and proper usage of API.
Feature discussions are the opposite - they are about properties of the game. For example damage of weapons, properties of blocks and so on. Such discussions should take place only in Issues.

**Comment by ClobberXD** [comment]
Please open a new issue if you'd like to discuss about feature-request approvals, as that's off-topic in this PR.

File: mods/ctf_crafting/depends.txt (hunk 2 above)
Comment (Panquesito7) [inline mods/ctf_crafting/depends.txt]: `depends.txt` is deprecated, use `mod.conf` instead.

File: mods/spears/depends.txt [hunk 4]
```diff
@@ -0,0 +1 @@
+default
```
Comment (Panquesito7) [inline mods/spears/depends.txt]: `depends.txt` is deprecated, use `mod.conf` instead.

**Comment by LoneWolfHT** [comment]
Closing for now. If you start working on this again let me know
